File: mailer/src/service.ts
import config from "./config.ts";
import type { Context, ISentInput } from "./context";
import * as tmpl from "./template.ts"

export const send = (
    ctx: Context,
    templateName: string,
    receiver: string,
    options: <string, string>
) => {
    if (templateName === "email_verification") {
        const template = tmpl.emailVerification(options["token"])


        ctx.mailgun.messages.create(config.mailgunDomain, {

        })

    }



}

File: mailer/service.go
package main

import (
	"context"
	"log/slog"
)

type Service struct {
	appURL      string
	frontendURL string

	mg *Mailgun
}

func NewService(appURL, frontendURL string, mg *Mailgun) *Service {
	return &Service{
		appURL:      appURL,
		frontendURL: frontendURL,
		mg:          mg,
	}
}

func (s Service) Send(
	ctx context.Context,
	cancel context.CancelFunc,
	receiver, templateName string,
	templateOpts map[string]string,
) error {
	tmpl, err := getTemplate(s.appURL, s.frontendURL, templateName)
	if err != nil {
		return err
	}

	t := tmpl(templateOpts)

	go func() {
		select {
		case <-ctx.Done():
			return
		default:
			if err := s.mg.Send(ctx, receiver, t.Subject, t.Body); err != nil {
				slog.ErrorContext(ctx, "failed to send email",
					"template_name", templateName,
					"err", err)
			}
			cancel()
		}
	}()

	return nil
}
